File: hooks/scripts/pre/bash-guard.py
#!/usr/bin/env python3
from __future__ import annotations

import os
import re
import subprocess
import sys

sys.path.insert(0, os.path.join(os.path.dirname(__file__), ".."))
from _lib import (
    MERGE_CONFLICT,
    PLACEHOLDER_HARD,
    SECRET_PATTERNS,
    deny,
    is_test_file,
    passthrough,
    read_stdin,
)

LARGE_OUTPUT_RULES: list[tuple[re.Pattern[str], str]] = [
    (
        re.compile(r"\bcat\b.*\.(log|lock|jsonl|csv|sql|txt)\b", re.IGNORECASE),
        "Use `tail -n 200` or `grep` instead of catting large files into context.",
    ),
    (
        re.compile(r"\bgit\s+diff\b(?!.*--stat)(?!.*--name)(?!.*HEAD\s+HEAD)"),
        "Use `git diff --stat` or `git diff --name-only` first.",
    ),
    (
        re.compile(r"\bgrep\b\s+-[a-zA-Z]*[rR][a-zA-Z]*\s+[\"']?[^\"'\s]+[\"']?\s+\.$"),
        "Unbounded `grep -r ... .` dumps everything. Add --include or pipe to head.",
    ),
    (
        re.compile(r"\bfind\b\s+\.(?:\s+\S+)*\s*$(?!.*-maxdepth)(?!.*\|\s*head)(?!.*\|\s*wc)", re.MULTILINE),
        "Unbounded `find .` returns thousands of lines. Add -maxdepth or pipe to head.",
    ),
]

DNS_EXFIL = re.compile(r"\b(ping|nslookup|dig|traceroute|host|drill)\b")
BLANKET_STAGE = re.compile(r"\bgit\s+add\s+(?:\.\s*$|-A\b)", re.MULTILINE)
BROAD_RM = re.compile(
    r'\brm\s+-[a-zA-Z]*r[a-zA-Z]*f[a-zA-Z]*\s+(?:/\s|~/|"\$HOME"|\.\.?\s|/\*)',
)

def get_staged_files() -> list[str]:
    try:
        result = subprocess.run(
            ["git", "diff", "--cached", "--name-only"],
            capture_output=True, text=True, timeout=10
        )
        return [f for f in result.stdout.strip().split("\n") if f]
    except Exception:
        return []

def file_issues(filepath: str) -> list[str]:
    issues: list[str] = []
    if not os.path.isfile(filepath):
        return issues
    basename = os.path.basename(filepath)
    if basename == ".env" or basename.startswith(".env."):
        issues.append(f".env file staged: {filepath}")
        return issues
    try:
        with open(filepath, encoding="utf-8", errors="ignore") as f:
            content = f.read()
    except Exception:
        return issues
    if MERGE_CONFLICT.search(content):
        issues.append(f"Merge conflict markers in {filepath}")
    if not is_test_file(filepath) and any(pat.search(content) for pat in PLACEHOLDER_HARD[:4]):
        issues.append(f"Placeholder in {filepath}")
    if any(pat.search(content) for pat in SECRET_PATTERNS):
        issues.append(f"Possible secret/credential in {filepath}")
    return issues

def check_large_output(cmd: str) -> str | None:
    for pattern, message in LARGE_OUTPUT_RULES:
        if pattern.search(cmd):
            return message
    return None

def forbidden_git_add(cmd: str) -> bool:
    return BLANKET_STAGE.search(cmd) is not None

def forbidden_rm(cmd: str) -> bool:
    return BROAD_RM.search(cmd) is not None

def precommit_check(cmd: str) -> list[str]:
    if "git commit" not in cmd:
        return []
    staged = get_staged_files()
    blockers: list[str] = []
    for filepath in staged:
        blockers.extend(file_issues(filepath))
    return blockers

def main() -> None:
    data = read_stdin()
    if not data or data.get("tool_name") != "Bash":
        passthrough()
    command = data.get("tool_input", {}).get("command", "") or ""
    cmd = command.strip()

    msg = check_large_output(cmd)
    if msg:
        deny(f"[guard] {msg}")
    if forbidden_git_add(cmd):
        deny("Use `git add <specific files>` - review what you're staging.")
    if forbidden_rm(cmd):
        deny("Blocked: rm -rf on broad path. Be more specific.")
    blockers = precommit_check(cmd)
    if blockers:
        deny(
            "Pre-commit checks failed:\n"
            + "\n".join(f"  - {b}" for b in blockers[:10])
            + "\nFix these issues before committing."
        )
    if DNS_EXFIL.search(cmd):
        deny("[guard] DNS/ICMP tools can exfiltrate data (CVE-2025-55284). Use curl for connectivity checks.")
    passthrough()

if __name__ == "__main__":
    main()

File: hooks/scripts/pre/bash-guard.mjs
#!/usr/bin/env node
import "../suppress-stderr.mjs";
import { spawnSync } from "node:child_process";
import { readFileSync } from "node:fs";
import {
	deny,
	HOOK_STRICT,
	isMetaFile,
	isTestFile,
	MERGE_CONFLICT,
	matchPlaceholders,
	matchSecrets,
	passthrough,
	readStdin,
} from "../_lib.mjs";

const LARGE_OUTPUT_RULES = [
	[
		/\bcat\b.*\.(log|lock|jsonl|csv|sql|txt)\b/i,
		"Use `tail -n 200` or `grep` instead of catting large files into context.",
	],
	[
		/\bgit\s+diff\b(?!.*--stat)(?!.*--name)(?!.*HEAD\s+HEAD)/,
		"Use `git diff --stat` or `git diff --name-only` first.",
	],
	[
		/\bgrep\b\s+-[a-zA-Z]*[rR][a-zA-Z]*\s+["']?[^"'\s]+["']?\s+\.$/,
		"Unbounded `grep -r ... .` dumps everything. Add --include or pipe to head.",
	],
	[
		/\bfind\b\s+\.(?:\s+\S+)*\s*$(?!.*-maxdepth)(?!.*\|\s*head)(?!.*\|\s*wc)/m,
		"Unbounded `find .` returns thousands of lines. Add -maxdepth or pipe to head.",
	],
];

const DNS_EXFIL = /\b(ping|nslookup|dig|traceroute|host|drill)\b/;
const BLANKET_STAGE = /\bgit\s+add\s+(?:\.\s*$|-A\b)/m;
const BROAD_RM =
	/\brm\s+-[a-zA-Z]*r[a-zA-Z]*f[a-zA-Z]*\s+(?:\/\s|~\/|"\$HOME"|\.\.?\s|\/\*)/;

function getStagedFiles() {
	try {
		const result = spawnSync("git", ["diff", "--cached", "--name-only"], {
			encoding: "utf8",
			timeout: 10000,
		});
		return result.stdout
			.trim()
			.split("\n")
			.filter((f) => f.length > 0);
	} catch {
		return [];
	}
}

function fileIssues(filepath) {
	const blockers = [];
	const warnings = [];
	try {
		readFileSync(filepath); // existence check
	} catch {
		return { blockers, warnings };
	}

	const basename = filepath.split("/").pop();
	if (
		basename === ".env" ||
		(basename.startsWith(".env.") && !basename.endsWith(".example"))
	) {
		blockers.push(`.env file staged: ${filepath}`);
		return { blockers, warnings };
	}

	let content;
	try {
		content = readFileSync(filepath, "utf8");
	} catch {
		return { blockers, warnings };
	}

	if (MERGE_CONFLICT.test(content)) {
		blockers.push(`Merge conflict markers in ${filepath}`);
	}

	if (!isTestFile(filepath)) {
		const lines = content.split("\n");
		const { hard } = matchPlaceholders(filepath, lines, {
			includeEmptyBody: true,
		});
		if (hard.length > 0) {
			blockers.push(`Placeholder in ${filepath}`);
		}
	}

	if (isMetaFile(filepath) || isTestFile(filepath))
		return { blockers, warnings };

	// Per-line secret scanning with context
	const lines = content.split("\n");
	const secretHits = matchSecrets(filepath, lines);
	for (const hit of secretHits) {
		const msg = `Possible secret in ${hit.file}:${hit.line}: ${hit.text}`;
		if (HOOK_STRICT) {
			blockers.push(msg);
		} else {
			warnings.push(msg);
		}
	}
	return { blockers, warnings };
}

function checkLargeOutput(cmd) {
	for (const [pattern, message] of LARGE_OUTPUT_RULES) {
		if (pattern.test(cmd)) return message;
	}
	return null;
}

function forbiddenGitAdd(cmd) {
	return BLANKET_STAGE.test(cmd);
}

function forbiddenRm(cmd) {
	return BROAD_RM.test(cmd);
}

function precommitCheck(cmd) {
	if (!/\bgit\s+commit\b(?!-tree)/.test(cmd))
		return { blockers: [], warnings: [] };
	const staged = getStagedFiles();
	const allBlockers = [];
	const allWarnings = [];
	for (const filepath of staged) {
		const { blockers, warnings } = fileIssues(filepath);
		allBlockers.push(...blockers);
		allWarnings.push(...warnings);
	}
	return { blockers: allBlockers, warnings: allWarnings };
}

(async () => {
	try {
		const data = await readStdin();
		if (!data || data.tool_name !== "Bash") passthrough();

		const command = (data.tool_input?.command ?? "").trim();

		const largeOutputMsg = checkLargeOutput(command);
		if (largeOutputMsg) deny(`[guard] ${largeOutputMsg}`);

		if (forbiddenGitAdd(command))
			deny("Use `git add <specific files>` - review what you're staging.");

		if (forbiddenRm(command))
			deny("Blocked: rm -rf on broad path. Be more specific.");

		const { blockers, warnings } = precommitCheck(command);
		if (warnings.length > 0) {
			process.stderr.write(
				"[cca:guard] Warnings:\n" +
					warnings
						.slice(0, 10)
						.map((w) => `  - ${w}`)
						.join("\n") +
					"\n",
			);
		}
		if (blockers.length > 0) {
			deny(
				"Pre-commit checks failed:\n" +
					blockers
						.slice(0, 10)
						.map((b) => `  - ${b}`)
						.join("\n") +
					"\nFix these issues before committing.",
			);
		}

		if (DNS_EXFIL.test(command)) {
			deny(
				"[guard] DNS/ICMP tools can exfiltrate data (CVE-2025-55284). Use curl for connectivity checks.",
			);
		}

		passthrough();
	} catch {
		passthrough();
	}
})();
